### PR TITLE
Proposer compare withdrawal roots

### PR DESCRIPTION
**Pull request by terencechain**
Proposer should compare local vs builder withdrawal roots, if they don't match then default to local block building. This prevents builder and relayer bugs result in block proposal failures

### CODE REVIEW

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go [hunk 1]
```diff
@@ -63,11 +63,28 @@ func (vs *Server) setExecutionData(ctx context.Context, blk interfaces.SignedBea
...
+				r1, err := localPayload.WithdrawalsRoot()
+				if err != nil {
+					return errors.Wrap(err, "failed to get local withdrawals root")
+				}
+				r2, err := builderPayload.WithdrawalsRoot()
+				if err != nil {
+					return errors.Wrap(err, "failed to get builder withdrawals root")
+				}
+				matchWithdrawalRoot := bytes.Equal(r1, r2)
+				if !matchWithdrawalRoot {
```
Comment (potuz) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go:76]:
```suggestion
				matchWithdrawalsRoot := bytes.Equal(r1, r2)
				if !matchWithdrawalsRoot {
```

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go [hunk 2]
```diff
@@ -63,11 +63,28 @@ func (vs *Server) setExecutionData(ctx context.Context, blk interfaces.SignedBea
...
+					log.WithFields(logrus.Fields{
+						"local":  r1,
+						"remote": r2,
+					}).Warn("Proposer: withdrawal roots don't match, using local")
```
Comment (potuz) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go:80]:
```suggestion
					}).Warn("Proposer: withdrawals roots don't match, using local")
```

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go [hunk 3]
```diff
@@ -63,11 +63,28 @@ func (vs *Server) setExecutionData(ctx context.Context, blk interfaces.SignedBea
...
+				}
+
 				// Compare payload values between local and builder. Default to the local value if it is higher.
 				builderValue, err := builderPayload.Value()
 				if err != nil {
 					log.WithError(err).Warn("Proposer: failed to get builder payload value") // Default to local if can't get builder value.
-				} else if builderValue.Cmp(localValue) > 0 {
+				} else if builderValue.Cmp(localValue) > 0 && matchWithdrawalRoot {
```
Comment (potuz) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go:87]:
```suggestion
				} else if builderValue.Cmp(localValue) > 0 && matchWithdrawalsRoot {
```

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go [hunk 4]
```diff
@@ -63,11 +63,28 @@ func (vs *Server) setExecutionData(ctx context.Context, blk interfaces.SignedBea
...
+				if err != nil {
+					return errors.Wrap(err, "failed to get builder withdrawals root")
+				}
+				matchWithdrawalRoot := bytes.Equal(r1, r2)
+				if !matchWithdrawalRoot {
+					log.WithFields(logrus.Fields{
+						"local":  r1,
+						"remote": r2,
+					}).Warn("Proposer: withdrawal roots don't match, using local")
+				}
```
Comment (potuz) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go:81]: An alternative is to add here an `} else` and prevent continuing if we already know that we are not going to use the builder cause the withdrawals root doesn't match

**Review by potuz — COMMENTED** [review]
LGTM

**Review by rauljordan — APPROVED** [review]

**Review by rauljordan — DISMISSED** [review]

**Review by potuz — APPROVED** [review]

**Review by saolyn — APPROVED** [review]

**Pull request merged**